### PR TITLE
Fix dirty dots not preserved when moving multiple cells

### PR DESCRIPTION
**Pull request by Alanhou1222**
## References
Fixes https://github.com/jupyterlab/jupyterlab/issues/16060
Closes https://github.com/jupyterlab/jupyterlab/pull/16161 (includes)

## Code changes
In packages/cells/src/model.ts, make the setDirty() function into a set function set isDirty().
In packages/notebook/src/widget.ts, move logics that set new cells dirty into the for loop of the moveCell function and make the code more clear.

## User-facing changes
dirty dots are kept even when moving multiple cells
image
image

## Backwards-incompatible changes
None

### CODE REVIEW

**Comment by Alanhou1222** [comment]
@krassowski Not sure if it's expected, but moved cells won't remember their content before being moved. Therefore, even if users revert the changes on a moved cell, the cell is still considered dirty.

**Comment by krassowski** [comment]
> but moved cells won't remember their content before being moved

Indeed, undo does not work after moving a cell - good catch! This is not expected, we should open a separate issue to track it and then fix it.

This PR looks good to me on the implementation side (the failing Python tests are unrelated); ideally we would also include unit tests in `packages/notebook/test/widget.spec.ts` for the dirty dot being preserved. Would you be interested in adding such a unit test (whether in this PR or in another)?

**Comment by Alanhou1222** [comment]
> > but moved cells won't remember their content before being moved
> 
> Indeed, undo does not work after moving a cell - good catch! This is not expected, we should open a separate issue to track it and then fix it.
> 
> This PR looks good to me on the implementation side (the failing Python tests are unrelated); ideally we would also include unit tests in `packages/notebook/test/widget.spec.ts` for the dirty dot being preserved. Would you be interested in adding such a unit test (whether in this PR or in another)?

Sure, I'll write some unit tests.

**Comment by Alanhou1222** [comment]
@krassowski I added two tests, one verifies dirty state is preserved after moving one code cell and another verifies dirty states are preserved after moving multiple cells. Let me know if you think there are other situations that should be tested.

File: packages/notebook/src/widget.ts [hunk 1]
```diff
@@ -418,6 +424,17 @@ export class StaticNotebook extends WindowedList<NotebookViewModel> {
...
+
+      if (from > to) {
+        if (this.widgets[to + i].model.type === 'code') {
+          (this.widgets[to + i].model as CodeCellModel).isDirty = dirtyState[i];
+        }
+      } else {
+        if (this.widgets[to + i - n + 1].model.type === 'code') {
+          (this.widgets[to + i - n + 1].model as CodeCellModel).isDirty =
+            dirtyState[i];
+        }
```
Comment (krassowski) [inline packages/notebook/src/widget.ts:436]: I wonder why is it that `rendered` and `headingCollapsed` do not need `- n + 1` but `isDirty` needs it? It seems a bit suspicious to me that the implementation here differs from the one for markdown cell properties. I suspect that one of those is wrong.

**Review by krassowski — APPROVED** [review]
Tested locally, works great. Thank you @Alanhou1222!

**Pull request merged**